### ISSUE TITLE
[Quantization] Enhance I2S quantization stability test with comprehensive statistical distributions

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#347]
## Problem Description

The `i2s_quantization_stability_test` in `crates/bitnet-models/src/gguf_min.rs` uses oversimplified test data generation that may not adequately validate quantization stability across real-world neural network weight distributions. The current implementation generates test data using a basic sinusoidal pattern: `((i as f32).sin() * scale_factor).clamp(-2.0, 2.0)`, which lacks the statistical diversity necessary for comprehensive quantization validation.

## Environment

- **File**: `crates/bitnet-models/src/gguf_min.rs`
- **Function**: `i2s_quantization_stability_test` (lines 781-802)
- **Quantization Type**: I2_S (2-bit signed quantization)
- **Current Pattern**: Simple sinusoidal wave with clamping
- **Test Framework**: Property-based testing with proptest

## Root Cause Analysis

**Current Implementation Limitations:**
1. **Limited Distribution Coverage**: Only sinusoidal patterns don't represent real neural network weight distributions
2. **Missing Edge Cases**: No testing of sparse, bimodal, or adversarial patterns common in quantized models
3. **Insufficient Validation**: Lacks coverage for distributions that stress quantization boundaries
4. **Production Gap**: Test patterns don't reflect actual model weight characteristics

**Technical Investigation:**
The current test data generation:
```rust
let mut test_data = vec![0.0f32; total_elems];
for (i, elem) in test_data.iter_mut().enumerate().take(total_elems) {
    *elem = ((i as f32).sin() * scale_factor).clamp(-2.0, 2.0);
}
```

This approach generates predictable, smooth patterns that may miss quantization instabilities that occur with:
- High-frequency patterns
- Sparse weight distributions 
- Outlier values near quantization boundaries
- Bimodal distributions common in trained neural networks

## Impact Assessment

**Severity**: Medium-High
**Affected Components**: 
- I2S quantization validation framework
- Cross-validation with C++ reference implementation
- Production model quantization reliability

**Risks**:
- Quantization bugs may not be detected during testing
- Production models might experience unexpected quantization artifacts
- Cross-validation accuracy may be compromised for certain weight distributions

### Primary Approach: Statistical Distribution Diversification

Enhance the test to include multiple statistical distributions that better represent real neural network scenarios:

```rust
#[test]
fn i2s_quantization_stability_test(
    block_count in 1usize..100,
    scale_factor in 0.1f32..10.0f32,
    distribution_type in prop_oneof![
        Just("uniform"),
        Just("normal"),
        Just("exponential"), 
        Just("bimodal"),
        Just("sparse"),
        Just("neural_weights"),
        Just("adversarial_high_freq"),
        Just("adversarial_boundaries"),
    ]
) {
    let layout = I2SLayout::default();
    let total_elems = block_count * layout.block_size;

    let test_data = match distribution_type {
        "uniform" => generate_uniform_data(total_elems, -1.0, 1.0),
        "normal" => generate_normal_data(total_elems, 0.0, 0.3),
        "exponential" => generate_exponential_data(total_elems, 1.0),
        "bimodal" => generate_bimodal_data(total_elems, -0.5, 0.5),
        "sparse" => generate_sparse_data(total_elems, 0.1), // 90% zeros
        "neural_weights" => generate_neural_weight_distribution(total_elems),
        "adversarial_high_freq" => generate_high_frequency_pattern(total_elems),
        "adversarial_boundaries" => generate_boundary_test_data(total_elems),
        _ => unreachable!(),
    };

    // Apply scale factor and clamp to valid range
    let test_data: Vec<f32> = test_data
        .into_iter()
        .map(|x| (x * scale_factor).clamp(-2.0, 2.0))
        .collect();

    // Existing validation logic
    let calculated_blocks = total_elems.div_ceil(layout.block_size);
    prop_assert_eq!(calculated_blocks, block_count);

    let required_bytes = calculated_blocks * layout.bytes_per_block;
    prop_assert!(required_bytes >= layout.bytes_per_block);
    prop_assert!(required_bytes <= total_elems * 4);
}
```

### Distribution Generator Functions

Create helper functions for each distribution type:

```rust
/// Generate uniform random data
fn generate_uniform_data(count: usize, min: f32, max: f32) -> Vec<f32> {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    (0..count).map(|_| rng.gen_range(min..max)).collect()
}

/// Generate normal distribution data (Box-Muller transform)
fn generate_normal_data(count: usize, mean: f32, std_dev: f32) -> Vec<f32> {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    (0..count)
        .map(|_| {
            let u1: f32 = rng.gen();
            let u2: f32 = rng.gen();
            let z0 = (-2.0 * u1.ln()).sqrt() * (2.0 * std::f32::consts::PI * u2).cos();
            mean + std_dev * z0
        })
        .collect()
}

/// Generate neural network weight-like distribution
fn generate_neural_weight_distribution(count: usize) -> Vec<f32> {
    // Combination of normal distribution with some sparse elements
    let mut data = generate_normal_data(count, 0.0, 0.2);
    // Make 20% of weights very small (pruned-like)
    for i in (0..count).step_by(5) {
        if i < data.len() {
            data[i] *= 0.01;
        }
    }
    data
}

/// Generate adversarial high-frequency pattern
fn generate_high_frequency_pattern(count: usize) -> Vec<f32> {
    (0..count)
        .map(|i| (i as f32 * 10.0).sin() * (i as f32 * 3.0).cos())
        .collect()
}

/// Generate boundary-testing data (values near quantization limits)
fn generate_boundary_test_data(count: usize) -> Vec<f32> {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    (0..count)
        .map(|i| {
            match i % 4 {
                0 => 1.99,  // Near positive clamp
                1 => -1.99, // Near negative clamp  
                2 => rng.gen_range(-0.1..0.1), // Near zero
                _ => rng.gen_range(-1.0..1.0),  // Normal range
            }
        })
        .collect()
}
```

### Phase 1: Core Enhancement (1-2 days)
- [ ] Add statistical distribution generator functions to test utilities
- [ ] Modify `i2s_quantization_stability_test` to use distribution selection
- [ ] Add `rand` crate dependency for statistical sampling (dev-dependencies)
- [ ] Ensure deterministic testing with seeded RNG

### Phase 2: Distribution Implementation (2-3 days)  
- [ ] Implement uniform distribution generator
- [ ] Implement normal distribution generator (Box-Muller)
- [ ] Implement neural weight distribution simulator
- [ ] Implement adversarial pattern generators
- [ ] Add sparse data generation (high zero percentage)

### Phase 3: Validation Enhancement (1-2 days)
- [ ] Add quantization accuracy validation for each distribution
- [ ] Implement round-trip error measurement
- [ ] Add boundary condition testing
- [ ] Ensure cross-validation compatibility

### Phase 4: Integration Testing (1 day)
- [ ] Run comprehensive test suite with new distributions
- [ ] Validate against existing C++ reference implementation
- [ ] Performance impact assessment
- [ ] Documentation updates

### Validation Criteria
1. **Distribution Coverage**: All 8 distribution types generate valid test data
2. **Quantization Stability**: No panics or invalid states for any distribution  
3. **Accuracy Preservation**: Round-trip error within acceptable bounds
4. **Performance**: Test execution time increase < 50%
5. **Determinism**: Reproducible results with seeded RNG

### Test Commands
```bash
# Run enhanced quantization stability tests
cargo test --no-default-features --features cpu i2s_quantization_stability_test

# Cross-validation with C++ reference
export BITNET_DETERMINISTIC=1 BITNET_SEED=42
cargo run -p xtask -- crossval

# Performance benchmarking
cargo bench --no-default-features --features cpu i2s_
```

## Related Issues

This issue is part of a broader effort to eliminate simulation-based implementations:

- **Related**: Issues addressing other simulation-based quantization tests
- **Related**: GPU quantization stability testing improvements  
- **Related**: Cross-validation framework enhancements
- **Depends on**: Statistical utilities for test data generation

## Acceptance Criteria

- [ ] Test uses 8+ distinct statistical distributions instead of sinusoidal pattern
- [ ] All distributions generate valid data within I2S quantization range
- [ ] Quantization stability validated across all distribution types
- [ ] Test execution remains deterministic with seeded RNG
- [ ] Cross-validation compatibility maintained
- [ ] Documentation updated with new test patterns
- [ ] Performance impact documented and acceptable (<50% slower)
- [ ] CI pipeline passes with enhanced tests

## Dependencies

- `rand` crate for statistical sampling (add to dev-dependencies)
- Proptest property combinators for distribution selection
- Existing I2S quantization infrastructure

## Risk Mitigation

- **Performance Impact**: Limit distribution complexity, use efficient generators
- **Test Flakiness**: Ensure deterministic seeding for all random generation
- **Compatibility**: Maintain existing test interface and validation logic
- **Complexity**: Start with core distributions, iterate based on findings

---

**Priority**: Medium-High  
**Effort**: 5-7 days  
**Labels**: quantization, testing, enhancement, simulation-cleanup

**Comment by EffortlessSteven** [comment]
Reviewed current codebase — this is already implemented. Closing as resolved.
